Implement Drag-n-Drop Functionality

GinaCastromonte opened the pull request:
### Description
This PR  implements drag-and-drop functionality for the family tree component, allowing users to rearrange family members by dragging their cards and dropping them onto other family members.

### Related Issues
- Create Cards Issue (https://github.com/Devocados/huellas-familiares/issues/2)

### Changes Made
  - Updated state management to handle the reassignment of children when a family member is dropped onto another.
  - Created the Family Card component to support drag-and-drop interactions.



### How to Test
**Prerequisites:**  
Ensure you run npm install. 

**Steps:**
1. **Set Up:** Run the app locally.
2. **Testing the Feature:**
   - Step 1: Drag a family member card and drop it onto another family member card.
   - Step 2: Confirm that the dropped family member is now a child of the target family member and that the connection lines are correctly displayed.

### Screenshots (if applicable)

![Screenshot 2024-10-11 at 3 27 15 PM](https://github.com/user-attachments/assets/55873700-f2d6-4326-b6b8-06483a98a8e4)


### Additional Notes
  - Still trying to integrate ' react-xarrows` to connect family members during dragging visually.  This will allow users to visually see the "family tree"

GinaCastromonte committed:
created simple form, started homepage

GinaCastromonte committed:
testing out drag and drop

GinaCastromonte committed:
cards seem functional

fsantamaria1 commented on components/forms/createMemberForm.tsx:
```diff
@@ -0,0 +1,64 @@
+            type="date"
+            id="memberDod"
+            className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
+            required
```
I don't think this field should be required

GinaCastromonte replied:
True!

fsantamaria1 commented on components/forms/createMemberForm.tsx:
```diff
@@ -0,0 +1,64 @@
+  return (
+    <div className="flex justify-center items-center min-h-screen">
+      <form className="max-w-sm w-full ">
+        <div className="mb-5">
```
Are we planning to ask the user for their name every time they create a new card?

fsantamaria1 replied:
I don't know what happened, but those were not the lines I selected lol

GinaCastromonte replied:
you're right, maybe a dropdown will be a better option.  I tried it out and will submit with this PR

fsantamaria1 replied:
Okayyyy

fsantamaria1 reviewed (changes requested):
@GinaCastromonte, please see my comments in the createMemberForm component

fsantamaria1 commented:
Also, if you `drag` a `card` and `don't drop` it on top of another one, you get this error:

![image](https://github.com/user-attachments/assets/271fa7a5-2c57-4def-bd20-9617b8dee4b2)

GinaCastromonte committed:
added dropdown, fixed requirement

GinaCastromonte commented:
1. I changed the form to have a dropdown for the current family members on the family tree. It allows the user to select the "parent member" when adding a new member to the family tree.  
2.  I deleted the DOD requirement from the form. 
3.  I couldn't recreate the error that you encountered with the drag functionality.  Could you please screenshot the console or any other error messages?

GinaCastromonte committed:
updated date on members

fsantamaria1 commented on components/forms/createMemberForm.tsx:
```diff
@@ -0,0 +1,318 @@
+    }
+  };
+
+  const getAllFamilyMembers = (members: FamilyMember): FamilyMember[] => {
```
We need an edge case here. Imagine that a user is creating relationships by dragging and dropping. If they create a relationship like this: Parent1 > Child1> Grandchild1 > Parent1 (again) this would basically create an infinite loop and give us a `runtime error`. 

![image](https://github.com/user-attachments/assets/e7471fc7-9b8f-4d69-9ff1-fc19dd19d786)

I am not really sure how we can prevent it 😅 but we need to have that edge case somewhere

fsantamaria1 replied:
Also, I swear I selected different lines lol

fsantamaria1 commented on components/familyCards/familyTree.tsx:
```diff
@@ -0,0 +1,271 @@
+    }
+    return null;
+  };
+  const getOrCreateRef = (id: string) => {
```
Eslint is complaining about this. It is sayng that since you are using a hook, it should be inside of a component. I am not sure what the best practice is when it comes to that since Typescript skills are not the best.

![image](https://github.com/user-attachments/assets/1acd1d9f-5698-4f41-81a2-650570585709)

fsantamaria1 replied:
Same thing here. Idk what's going on with my line selection

GinaCastromonte committed:
fixed error

GinaCastromonte committed:
ES Lint issue

GinaCastromonte committed:
ESLint error

fsantamaria1 commented on components/forms/createMemberForm.tsx:
```diff
@@ -0,0 +1,318 @@
+            required
+          />
+        </div>
+        \{" "}
```
I think you forgot to remove this slash

fsantamaria1 reviewed (changes requested):
I tested it and the infinite loop doesn't happen anymore. That's AWESOME!
I guess ESLint is still mad about something

GinaCastromonte committed:
eslint issue
